### ISSUE TITLE
Allow multiple assignment (i.e. tuple on LHS) in walrus operator

### ISSUE THREAD

**Issue by 02ad864c-ae2a-4ade-b93d-57afad65b191** [python/cpython#87309]
BPO | [43143](https://bugs.python.org/issue43143)
--- | :---
Nosy | @gvanrossum, @tim-one, @rhettinger, @pfalcon, @emilyemorehouse, @lysnikolaou

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2021-02-06.08:13:07.477>
labels = ['interpreter-core', '3.10']
title = 'Allow multiple assignment (i.e. tuple on LHS) in walrus operator'
updated_at = <Date 2021-02-08.07:31:26.572>
user = 'https://github.com/pfalcon'
```

bugs.python.org fields:
```python
activity = <Date 2021-02-08.07:31:26.572>
actor = 'pfalcon'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Interpreter Core']
creation = <Date 2021-02-06.08:13:07.477>
creator = 'pfalcon'
dependencies = []
files = []
hgrepos = []
issue_num = 43143
keywords = []
message_count = 3.0
messages = ['386551', '386564', '386620']
nosy_count = 6.0
nosy_names = ['gvanrossum', 'tim.peters', 'rhettinger', 'pfalcon', 'emilyemorehouse', 'lys.nikolaou']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = None
url = 'https://bugs.python.org/issue43143'
versions = ['Python 3.10']
```

</p></details>

**Comment by andylamp** [comment]
Just commenting there, one of the major benefits of the walrus operator is to make if cases with assignment and checking easier and more readable.

Let us consider this case,

```python
if ((a, b) := foo) and (a > 1 or b < 2):
    # do stuff
```

Currently statements like that fail, which is a shame... is it still not considered a valid use case? If so, are there any suggestions to make such cases prettier and more compact?

Thanks!

**Comment by KaiPetzke** [comment]
The use case suggested by @andylamp can currently be written as follows:

    if (tuple := foo()) and ((a := tuple[0]) > 1 or (b := tuple[1]) < 2):
        # do stuff

The Python team has been slow to accept the walrus operator into the language to avoid ambiguities and hard to read source code. But ambiguities will directly arise, when the left- and right-hand side of the walrus operator are of different type in a compound assignment expression. For example, should the following call print `ab` according to the right-hand side, which  is a two-byte string, or `("a", "b")` according to the re-assembled left-hand side, which would be a two-element tuple of one-byte strings:

    print(((a, b) := "ab"))

And what would the next statement print? `(1, 2, 3, 4)` from the right-hand side or a piece-wise reconstructed left-hand side, or `(4, 4, 3, 4)` from reconstructing the left-hand side only after the assignment is complete?

    print(((dummy, dummy, x, dummy) := (1, 2, 3, 4)))

And things become even more complex, when iterators with side effects come into the show.

**Comment by andylamp** [comment]
@KaiPetzke yep, I am aware this can be achieved that way; however it is a bit verbose compared to the vanilla walrus one.

I also understand the difficulties integrating the walrus operator in the base language, but wanted to flag this valid use-case.

**Comment by huili80** [comment]
> The use case suggested by @andylamp can currently be written as follows:
> 
> ```
> if (tuple := foo()) and ((a := tuple[0]) > 1 or (b := tuple[1]) < 2):
>     # do stuff
> ```
> 
> The Python team has been slow to accept the walrus operator into the language to avoid ambiguities and hard to read source code. But ambiguities will directly arise, when the left- and right-hand side of the walrus operator are of different type in a compound assignment expression. For example, should the following call print `ab` according to the right-hand side, which is a two-byte string, or `("a", "b")` according to the re-assembled left-hand side, which would be a two-element tuple of one-byte strings:
> 
> ```
> print(((a, b) := "ab"))
> ```
> 
> And what would the next statement print? `(1, 2, 3, 4)` from the right-hand side or a piece-wise reconstructed left-hand side, or `(4, 4, 3, 4)` from reconstructing the left-hand side only after the assignment is complete?
> 
> ```
> print(((dummy, dummy, x, dummy) := (1, 2, 3, 4)))
> ```
> 
> And things become even more complex, when iterators with side effects come into the show.

None of these are unique to having a tuple on the left-hand side, consider the following example
```
print((a := (a := 1, a + 1)))
```
of which the result is
```(1, 2)```

So this example makes it clear that
1. The result of an assignement expression is the expression on the right-hand side, which is actually stated in the doc (https://docs.python.org/3/reference/expressions.html#grammar-token-python-grammar-assignment_expression).
2. The expression is evaluated before the assignment, and is *not* re-evaluated after the assignment, which obviously makes sense.

Given the above observed behavior, I'm not sure there would be ambiguity at all that could result from having a tuple on the left-hand side of the walrus operator.

**Comment by p4l1ly** [comment]
This would also enable more descriptive (and maybe more performing) `let` statements in list comprehensions. With the following definition,

```
def let(x):
  return True
```

one could write

```
xs = [x for ab in abs if let((x,y) := ab2xy(ab)) if y]
```

instead of

```
xs = [x for ab in abs for x, y in (ab2xy(ab),) if y]
```

**Comment by KaiPetzke** [comment]
> one could write
> 
> ```
> xs = [x for ab in abs if let((x,y) := ab2xy(ab)) if y]
> ```

One can already write today:

```
xs = [ab2y_res[0] for ab in abs if (ab2y_res := ab2y(ab)) and ab2y_res[1]]
```

It is clearly a bit more verbose than the walrus double assignment. On the other hand, assigning the  result of `ab2y()` to just one variable and then indexing it, will likely be more readable to someone, who has to read someone else's code, as only one new temporary variable is involved instead of two.

**Comment by p4l1ly** [comment]
I don't agree. Variables with reasonable names are more readable than positional indexing. Also, in a more complex case, one loses the main idea of the code in a sea of brackets.

**Comment by Root-Core** [comment]
An absolute stupid workaround, but it's working:

```python
def create_tuple(i: int) -> Tuple[str, str, dict]:
    return f"One_{i}", f"Two_{i}", { "z": f"Three_{i}" }

records = [
    {
        "x": x,
        "y": y,
    } | z
    for index in range(5)
    for x, y, z in [create_tuple(index)]
]
```

Just wrap it in a list and iterate over it. This is a waste of CPU cycles.. but not too bad.
It would be nice, if walrus operator unpacking were a thing.
